### PR TITLE
Api-server: Add parallelism for rbac storageProvider postStartHook

### LINKED ISSUE #97119: Optimize rbac/bootstrap-roles post-startup-hook
The [rbac/bootstrap-roles post startup-hook](https://github.com/kubernetes/kubernetes/blob/9bf7fcfc76eb809ac74ce1f5f36f95de5e3a06ba/pkg/registry/rbac/rest/storage_rbac.go#L141) is slooooow. 

**Why:** It needs to reconcile ~100 roles/clusterroles/rolebings/clusterrolebindings, each of them requiring 1-2 api calls (get + optional update). It does all operations serially and in case of any error it starts from the beginning. In case of overloaded api-server it may legitimately take O(minutes) for this post-startup hook to complete.

**Why important:** Api-server won't become healthy/ready as long as all post-startup-hook completes. As a result, an overloaded kube-apiserver that was restarted may require O(minutes) to become healthy again. This may cause unexpected issues if the liveness probe is not configured to allow that, e.g. it may end up in kube-apiserver going into restart loop. 

**What would you like to be added**:
Luckily, there are two low-hanging-fruits that should allow to significantly speed up:

- [ ] Increase parallelism - while we cannot execute all api-call in parallel (some may depend on each other, e.g. creating role before role-binding) we can definitely do better than executing everything serially
- [ ] Fix error handling to not start from the beginning - we could resume from the failed operations

/sig api-machinery
/sig scalability

### PR DESCRIPTION
**Pull request by gavinfish**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Parallelize start-post-hook of rbac RestStorageProvider. This can decrease api-server starting time. Part of #97119.

**Which issue(s) this PR fixes**:

Fix #97119

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
none
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
none
```

### CODE REVIEW

File: pkg/registry/rbac/rest/storage_rbac.go [hunk 1]
```diff
@@ -238,9 +245,11 @@ func (p *PolicyData) EnsureRBACPolicy() genericapiserver.PostStartHookFunc {
 					failedReconciliation = true
```
Comment (liggitt) [inline pkg/registry/rbac/rest/storage_rbac.go:245]: writing to closure state here introduces a data race (same comment applies to all)

**Review by liggitt — COMMENTED** [review]
Let's also include tracking of which objects were successfully reconciled and skip them on retry

**Comment by liggitt** [comment]
are these operations limited by QPS/burst settings on the loopback client?

**Comment by gavinfish** [comment]
> are these operations limited by QPS/burst settings on the loopback client?

@liggitt  I think so. In this case, do we need a dynamic value from loopback config or default one is ok for parallelism.

File: pkg/registry/rbac/rest/storage_rbac.go [hunk 2]
```diff
@@ -171,10 +177,16 @@ func retryOnConflictOrServiceUnavailable(backoff wait.Backoff, fn func() error)
 
 func (p *PolicyData) EnsureRBACPolicy() genericapiserver.PostStartHookFunc {
 	return func(hookContext genericapiserver.PostStartHookContext) error {
+		ctx := context.Background()
+		visitedClusterRoleSet := rbacregistry.NewThreadSafeSet()
+		visitedClusterRoleBindingSet := rbacregistry.NewThreadSafeSet()
+		visitedRoleSet := rbacregistry.NewThreadSafeSet()
+		visitedRoleBindingSet := rbacregistry.NewThreadSafeSet()
```
Comment (liggitt) [inline pkg/registry/rbac/rest/storage_rbac.go:184]: nit, rename `visited` to `completed`

File: pkg/registry/rbac/rest/storage_rbac.go [hunk 3]
```diff
@@ -65,7 +68,10 @@ type RESTStorageProvider struct {
 	Authorizer authorizer.Authorizer
 }
 
-var _ genericapiserver.PostStartHookProvider = RESTStorageProvider{}
+var (
+	_           genericapiserver.PostStartHookProvider = RESTStorageProvider{}
+	parallelism                                        = 5
```
Comment (liggitt) [inline pkg/registry/rbac/rest/storage_rbac.go:73]: how was this value selected? what was the impact of lower and higher values on startup time (or a value of `1` matching the current behavior)?

File: pkg/registry/rbac/rest/storage_rbac.go [hunk 4]
```diff
@@ -65,7 +68,10 @@ type RESTStorageProvider struct {
 	Authorizer authorizer.Authorizer
 }
 
-var _ genericapiserver.PostStartHookProvider = RESTStorageProvider{}
+var (
+	_           genericapiserver.PostStartHookProvider = RESTStorageProvider{}
```
Comment (liggitt) [inline pkg/registry/rbac/rest/storage_rbac.go:72]: nit, add a newline to avoid the weird whitespace

**Comment by liggitt** [comment]
for PRs that make performance-related changes, it's helpful to include before/after comparisons in the description to help evaluate the benefit vs the added code complexity

**Comment by mm4tt** [comment]
> for PRs that make performance-related changes, it's helpful to include before/after comparisons in the description to help evaluate the benefit vs the added code complexity
+1

Can we start with writing a benchmark for the existing implementation? You could create a benchmark that mocks the k8sclient with the [fake client](https://github.com/kubernetes/kubernetes/blob/9bf7fcfc76eb809ac74ce1f5f36f95de5e3a06ba/pkg/registry/rbac/rest/storage_rbac.go#L185) and measures the execution time of the `EnsureRBACPolicy()` method. It should be relatively simple and can be done it a separate PR. Once we have it we'll have a clear signal of how much this implementation speeds things up and it would also help us to choose the proper `parallelism` value. 

Anyway, thanks so much for taking on this one!

**Comment by gavinfish** [comment]
> > for PRs that make performance-related changes, it's helpful to include before/after comparisons in the description to help evaluate the benefit vs the added code complexity
> > +1
> 
> Can we start with writing a benchmark for the existing implementation? You could create a benchmark that mocks the k8sclient with the [fake client](https://github.com/kubernetes/kubernetes/blob/9bf7fcfc76eb809ac74ce1f5f36f95de5e3a06ba/pkg/registry/rbac/rest/storage_rbac.go#L185) and measures the execution time of the `EnsureRBACPolicy()` method. It should be relatively simple and can be done it a separate PR. Once we have it we'll have a clear signal of how much this implementation speeds things up and it would also help us to choose the proper `parallelism` value.
> 
> Anyway, thanks so much for taking on this one!

Sure, let me have a try. Wait a for a benchmark before we go on this thread.

/hold

**Comment by mm4tt** [comment]
Hey, @gavinfish. Sorry for not being responsive, I was OOO for the christmas / new year holidays. 

I saw that you implemented the benchmark, it's great. Could you please rebase this PR and provide benchmark results with and without this change? Thanks!

**Comment by gavinfish** [comment]
@mm4tt I have done some tests with or without this change. It seems that code without the parallelism has better performance. But I think there maybe a gap between the real reconciling roles time between the fake one. Only if the works take some time, the parallelism framework can show its power.

|parallelism|name|times|speed|
|--|--|--|--|
|NONE|BenchmarkEnsureRBACPolicy-4   |       408   |        2856402 ns/op| 
|1|BenchmarkEnsureRBACPolicy-4         | 373        |    3194145 ns/op | 
|2|BenchmarkEnsureRBACPolicy-4        |  316        |    3708861 ns/op   | 
|3|BenchmarkEnsureRBACPolicy-4         | 310        |    3777283 ns/op     | 
|4|BenchmarkEnsureRBACPolicy-4         | 308       |     3802364 ns/op   | 
|5|BenchmarkEnsureRBACPolicy-4         | 292       | 4142317 ns/op   |

**Comment by mm4tt** [comment]
I believe this is kind of expected given that you used a plain fake client. It serves the requests from memory without any blocking so the parallelism  makes things rather worse than better because of the overhead cost of goroutine scheduling and synchronization. I'd recommend to change the benchmark to simulate some network latency, e.g. 100ms per request which is a reasonable approximation of moderately loaded api-server, and to re-run the benchmarks. Below an example code snippets with the benchmark modifycation:

```
func BenchmarkEnsureRBACPolicy(b *testing.B) {
	var policy = &PolicyData{
		ClusterRoles:               append(bootstrappolicy.ClusterRoles(), bootstrappolicy.ControllerRoles()...),
		ClusterRoleBindings:        append(bootstrappolicy.ClusterRoleBindings(), bootstrappolicy.ControllerRoleBindings()...),
		Roles:                      bootstrappolicy.NamespaceRoles(),
		RoleBindings:               bootstrappolicy.NamespaceRoleBindings(),
		ClusterRolesToAggregate:    bootstrappolicy.ClusterRolesToAggregate(),
		ClusterRoleBindingsToSplit: bootstrappolicy.ClusterRoleBindingsToSplit(),
	}
	for n := 0; n < b.N; n++ {
		coreClientSet := fake.NewSimpleClientset()
		coreClientSet.PrependReactor("*", "*", func(cgtesting.Action) (bool, runtime.Object, error) {
			time.Sleep(50 * time.Millisecond)
			return false, nil, nil
		})
		_, _ = ensureRBACPolicy(policy, coreClientSet)
	}
}
```

**Comment by gavinfish** [comment]
Thanks @mm4tt  for your suggestion. With the  simulation change, I have just realized that the client is indeed shared between the parallelism workers now. And for the reconcile job, it should spend most time on client get/create functions. So simply parallelism will not help here.

https://github.com/kubernetes/kubernetes/blob/564b0e55c7007745500d579356897848aaacb9dd/staging/src/k8s.io/component-helpers/auth/rbac/reconciliation/reconcile_role.go#L144-L153

**Comment by mm4tt** [comment]
> Thanks Matt Matejczyk for your suggestion. With the simulation change, I have just realized that the client is indeed shared between the parallelism workers now. And for the reconcile job, it should spend most time on client get/create functions. So simply parallelism will not help here.
> 
> https://github.com/kubernetes/kubernetes/blob/564b0e55c7007745500d579356897848aaacb9dd/staging/src/k8s.io/component-helpers/auth/rbac/reconciliation/reconcile_role.go#L144-L153

Not sure If I understand, would you mind elaborating more on it?

**Comment by mm4tt** [comment]
Hey @gavinfish, could you explain why do you believe parallelism want help here? I believe it should be quite the opposite as the rbac-bootstrap spends majority of time time on calling api and waiting for the responses. Please let me know, I believe we're actually pretty close to making it visibly faster.

**Comment by gavinfish** [comment]
@mm4tt Just ignore my last comment. I have just uploaded the latest code changes so maybe you can review it and help me solve the problem which confuses me now. With current changes, the benchmarks are really close when I change the parallelism variable. There should be something crucial I missed.

**Comment by mm4tt** [comment]
Sure, could you share the benchmark results with different parallelism setting as you did last time?

**Comment by gavinfish** [comment]
@mm4tt sure.

|parallelism|name|times|speed|
|--|--|--|--|
|NONE|BenchmarkEnsureRBACPolicy-4   |       1   |        13373106905 ns/op| 
|1|BenchmarkEnsureRBACPolicy-4         | 1        |    13488402720 ns/op | 
|2|BenchmarkEnsureRBACPolicy-4        |  1        |    13531308796 ns/op   | 
|4|BenchmarkEnsureRBACPolicy-4         | 1        |    13522369803 ns/op     | 
|8|BenchmarkEnsureRBACPolicy-4         | 1       |     13509797863 ns/op   | 
|16|BenchmarkEnsureRBACPolicy-4         | 1       | 13549481207 ns/op   |

**Comment by liggitt** [comment]
I think the parallelism and skipping retry of successfully reconciled objects is useful to minimize the startup time and failure rate. I'd like to see this completed if you can rebase it and resolve the question about the client-side rate limiting (https://github.com/kubernetes/kubernetes/pull/97123#issuecomment-740970087)

**Pull request closed without merging**